Add forceDeploy property in Cognito User Pools

TsimpDim opened the pull request:
I tried to implement the proposed solution which can be found [here](https://github.com/serverless/serverless/issues/9635#issuecomment-874320921). I added a `forceDeploy` flag that once set will set a random property in the custom resource that is used to attach triggers - if the flag is set to false then it sets the property to `0`. I updated tests as well, and made sure the linter is happy.


Addresses: #9635

medikoo commented on lib/plugins/aws/package/compile/events/cognitoUserPool.js:
```diff
@@ -171,6 +172,7 @@ class AwsCompileCognitoUserPoolEvents {
             }
 
             let customCognitoUserPoolResource;
+            const forceDeployProperty = forceDeploy ? Math.random() * 10 : 0;
```
Let's maybe set it to `Date.now()`, and do not set `ForceDeploy` (we can assign `undefined`) if no `forceDeploy` was set

TsimpDim replied:
Updated!

medikoo reviewed (changes requested):
Thank you @TsimpDim, that looks very promising. Have you confirmed locally it does the job?

TsimpDim commented:
> Have you confirmed locally it does the job?

During deployment the Triggers still get removed for a little while until CF updates all the resources, but they eventually get added back. I'm not sure if we could improve this at all, but either way it is a way better solution then what one has to do to circumvent this issue currently :D

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/cognitoUserPool.test.js:
```diff
@@ -339,6 +339,77 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
+    });
+
+    it('should create the necessary resources for the most minimal configuration with forceDeploy', () => {
+      awsCompileCognitoUserPoolEvents.serverless.service.functions = {
```
Sorry, didn't notice that one earlier.

In new tests we always rely on `runServerless` util (we have already a tons of tests configured for it, which can self as reference).

Can you update that one, so it's also `runServerless` based?

TsimpDim replied:
Sure! I had to do some reading because I'm not familiar at all with testing in Javascript (or the internals of Serverless) but I came up with this:

```
    it('should create the necessary resources for the most minimal configuration with forceDeploy', () =>
      runServerless({
        fixture: 'cognitoUserPool',
        configExt: {
          functions: {
            existingSimple: {
              events: [
                {
                  cognitoUserPool: {
                    forceDeploy: true
                  },
                },
              ],
            }
          }
        },
        command: 'package',
      }).then(({ awsNaming, cfTemplate }) => {
        const { Resources } = cfTemplate;
        let customResource = Resources[awsNaming.getCustomResourceCognitoUserPoolResourceLogicalId('existingSimple')];
        
        expect(customResource.Properties.ForceDeploy).to.not.deep.equal(undefined);
    }));
```

(only `expect()` statement instead of the previous three). Will something like this do?

medikoo replied:
Yes, that looks very good.
Just use async/await (instead of `.then()`), and rather confirm on property as `expect(typeof ..).to.equal('string')`

TsimpDim commented:
I updated the test: 

I could not find a better way of specifying what `ForceDeploy`'s value should be (especially since it varies from execution to execution) so I opted for a value range assessment. 
I also tried to add 
`expect(addCustomResourceToServiceStub).to.have.been.calledOnce;` which I would expect to run fine, but it always failed (`0` calls instead of the expected one - not sure if that's expected or I did something wrong)

I also updated the documentation to show the new flag as well.

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/cognitoUserPool.test.js:
```diff
@@ -339,11 +340,41 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
+      const customResource =
+        Resources[awsNaming.getCustomResourceCognitoUserPoolResourceLogicalId('existingSimple')];
+
+      expect(customResource.Properties.ForceDeploy).to.not.deep.equal(undefined);
+      expect(customResource.Properties.ForceDeploy).to.be.below(Date.now()).and.to.be.above(0);
```
Let's simplify it to:

```javascript
expect(typeof customResource.Properties.ForceDeploy).to.equal('number')
```

TsimpDim replied:
Fixed

medikoo commented on test/unit/lib/plugins/aws/package/compile/events/cognitoUserPool.test.js:
```diff
@@ -339,11 +340,41 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
       });
     });
 
+    it('should create the necessary resources for the most minimal configuration with forceDeploy', async () => {
```
Let's rename it to ``should support `forceDeploy` setting``

TsimpDim replied:
Fixed

medikoo reviewed (commented):
Thank you @TsimpDim I have just two minor suggestions, and otherwise, it looks good to me

medikoo reviewed (approved):
Thank you @TsimpDim that looks great!

meirgbinahai commented:
I see that this feature was pushed to version [2.71.0](https://github.com/serverless/serverless/releases/tag/v2.71.0) - can anyone confirm that it works in versions `2.71.0+` ?

chris commented:
One followup question, related to @TsimpDim 's comment about how the triggers seem to get removed for a little bit during the deploy, then added back in. If that's the case, it seems like unless you change your Cognito pool frequently, it'd be best to only add the `forceDeploy: true` part in when you know this issue will happen? i.e. if we are only rarely changing user pool config (so this issue would rarely occur), I wouldn't want to have the triggers disappear on every deploy, and risk problems in my system (e.g. triggers not getting called) for the more common deploy case. Does that make sense - i.e. would you agree that if you infrequently change your user pool, you should only use `forceDeploy: true` when making those occasional changes, and otherwise leave it out of serverless.yml?

AEsmerio commented:
I have been struggling with this issue for a very long time, but finally been resolved! THANKS! By adding "forceDeploy: true" it solved the problem. 🍻
